File: backend/src/modules/placement/placement.repository.ts
import { prisma } from '../../config/database';
import { ApplicationStatus, DriveStatus } from '@prisma/client';
import {
  CreateDriveDto,
  ApplyDriveDto,
  UpdateApplicationStatusDto,
  ScheduleInterviewDto,
  RespondOfferDto,
  DriveQueryDto,
} from './placement.types';

export class PlacementRepository {
  // Placement Officer Dashboard & Metrics
  async getOfficerDashboardStats(collegeId: string) {
    const [totalDrives, activeDrives, totalApplications, offeredCount, acceptedOffers] = await Promise.all([
      prisma.placementDrive.count({ where: { college_id: collegeId } }),
      prisma.placementDrive.count({ where: { college_id: collegeId, status: DriveStatus.ONGOING } }),
      prisma.placementApplication.count({ where: { drive: { college_id: collegeId } } }),
      prisma.placementApplication.count({
        where: { drive: { college_id: collegeId }, status: ApplicationStatus.OFFERED },
      }),
      prisma.placementApplication.count({
        where: { drive: { college_id: collegeId }, offer_status: 'ACCEPTED' },
      }),
    ]);

    return {
      totalDrives,
      activeDrives,
      totalApplications,
      offeredCount,
      acceptedOffers,
    };
  }

  // Student Dashboard & Applications Tracking
  async getStudentDashboardStats(userId: string) {
    const [myApplications, myOffers, upcomingInterviews] = await Promise.all([
      prisma.placementApplication.findMany({
        where: { student_id: userId },
        include: {
          drive: true,
          interviews: { orderBy: { scheduled_at: 'asc' } },
        },
        orderBy: { applied_at: 'desc' },
      }),
      prisma.placementApplication.findMany({
        where: { student_id: userId, status: ApplicationStatus.OFFERED },
        include: { drive: true },
      }),
      prisma.placementInterview.findMany({
        where: { application: { student_id: userId }, scheduled_at: { gte: new Date() } },
        include: { application: { include: { drive: true } } },
        orderBy: { scheduled_at: 'asc' },
      }),
    ]);

    return {
      myApplications,
      myOffers,
      upcomingInterviews,
    };
  }

  // Drives
  async createDrive(collegeId: string, dto: CreateDriveDto) {
    return prisma.placementDrive.create({
      data: {
        college_id: collegeId,
        company_name: dto.company_name,
        role_title: dto.role_title,
        package_ctc: dto.package_ctc,
        location: dto.location,
        eligibility: dto.eligibility,
        min_cgpa: dto.min_cgpa || 0.0,
        allowed_departments: dto.allowed_departments || [],
        max_backlogs: dto.max_backlogs || 0,
        job_description: dto.job_description,
        deadline: new Date(dto.deadline),
        status: dto.status || DriveStatus.UPCOMING,
      },
      include: {
        _count: { select: { applications: true } },
      },
    });
  }

  async findDrives(collegeId: string, query: DriveQueryDto) {
    const page = query.page || 1;
    const limit = query.limit || 20;
    const skip = (page - 1) * limit;

    const where: Record<string, unknown> = { college_id: collegeId };
    if (query.status) where.status = query.status;
    if (query.search) {
      where.OR = [
        { company_name: { contains: query.search, mode: 'insensitive' } },
        { role_title: { contains: query.search, mode: 'insensitive' } },
      ];
    }

    const [total, drives] = await Promise.all([
      prisma.placementDrive.count({ where: where as never }),
      prisma.placementDrive.findMany({
        where: where as never,
        skip,
        take: limit,
        orderBy: { created_at: 'desc' },
        include: {
          _count: { select: { applications: true } },
        },
      }),
    ]);

    return { total, page, limit, drives };
  }

  async findDriveById(driveId: string) {
    return prisma.placementDrive.findUnique({
      where: { id: driveId },
      include: {
        applications: {
          include: {
            student: {
              select: {
                id: true,
                first_name: true,
                last_name: true,
                email: true,
                avatar_url: true,
                department: { select: { id: true, code: true, name: true } },
              },
            },
            interviews: true,
          },
        },
        _count: { select: { applications: true } },
      },
    });
  }

  // Applications
  async createApplication(studentId: string, dto: ApplyDriveDto) {
    return prisma.placementApplication.create({
      data: {
        drive_id: dto.drive_id,
        student_id: studentId,
        resume_url: dto.resume_url,
      },
      include: {
        drive: true,
      },
    });
  }

  async findApplication(studentId: string, driveId: string) {
    return prisma.placementApplication.findUnique({
      where: { drive_id_student_id: { drive_id: driveId, student_id: studentId } },
    });
  }

  async updateApplicationStatus(applicationId: string, dto: UpdateApplicationStatusDto) {
    return prisma.placementApplication.update({
      where: { id: applicationId },
      data: {
        status: dto.status,
        offer_ctc: dto.offer_ctc,
        offer_status: dto.status === ApplicationStatus.OFFERED ? 'PENDING' : undefined,
      },
      include: {
        drive: true,
        student: { select: { id: true, first_name: true, last_name: true, email: true } },
      },
    });
  }

  // Interviews & Offers
  async scheduleInterview(dto: ScheduleInterviewDto) {
    const interview = await prisma.placementInterview.create({
      data: {
        application_id: dto.application_id,
        round_name: dto.round_name,
        scheduled_at: new Date(dto.scheduled_at),
        location_or_link: dto.location_or_link,
        notes: dto.notes,
      },
    });

    await prisma.placementApplication.update({
      where: { id: dto.application_id },
      data: { status: ApplicationStatus.INTERVIEW_SCHEDULED },
    });

    return interview;
  }

  async respondToOffer(applicationId: string, studentId: string, dto: RespondOfferDto) {
    return prisma.placementApplication.update({
      where: { id: applicationId, student_id: studentId, status: ApplicationStatus.OFFERED },
      data: {
        offer_status: dto.offer_status,
      },
      include: { drive: true },
    });
  }
}

File: apps/campus_hub_app/lib/features/placement/data/placement_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../../core/network/api_client.dart';
import '../domain/placement_models.dart';

class PlacementRepository {
  final Dio _dio;

  PlacementRepository(this._dio);

  Future<Map<String, dynamic>> getOfficerDashboardStats() async {
    final response = await _dio.get('/api/v1/placement/dashboard/officer');
    return response.data['data'] ?? {};
  }

  Future<Map<String, dynamic>> getStudentDashboardStats() async {
    final response = await _dio.get('/api/v1/placement/dashboard/student');
    return response.data['data'] ?? {};
  }

  Future<List<PlacementDriveModel>> getDrives({String? status, String? search}) async {
    final query = <String, dynamic>{};
    if (status != null && status.isNotEmpty) query['status'] = status;
    if (search != null && search.isNotEmpty) query['search'] = search;

    final response = await _dio.get('/api/v1/placement/drives', queryParameters: query);
    final data = response.data['data'];
    final List list = data is Map ? (data['drives'] ?? []) : (data ?? []);
    return list.map((json) => PlacementDriveModel.fromJson(json)).toList();
  }

  Future<PlacementDriveModel> getDriveDetails(String id) async {
    final response = await _dio.get('/api/v1/placement/drives/$id');
    return PlacementDriveModel.fromJson(response.data['data']);
  }

  Future<PlacementDriveModel> createDrive({
    required String companyName,
    required String roleTitle,
    required String deadline,
    String? packageCtc,
    String? location,
    String? eligibility,
    double? minCgpa,
    List<String>? allowedDepartments,
    int? maxBacklogs,
    String? jobDescription,
  }) async {
    final response = await _dio.post(
      '/api/v1/placement/drives',
      data: {
        'company_name': companyName,
        'role_title': roleTitle,
        'deadline': deadline,
        'package_ctc': packageCtc,
        'location': location,
        'eligibility': eligibility,
        'min_cgpa': minCgpa,
        'allowed_departments': allowedDepartments,
        'max_backlogs': maxBacklogs,
        'job_description': jobDescription,
      },
    );
    return PlacementDriveModel.fromJson(response.data['data']);
  }

  Future<PlacementApplicationModel> applyForDrive(String driveId, {String? resumeUrl}) async {
    final response = await _dio.post(
      '/api/v1/placement/apply',
      data: {
        'drive_id': driveId,
        'resume_url': resumeUrl,
      },
    );
    return PlacementApplicationModel.fromJson(response.data['data']);
  }

  Future<void> updateApplicationStatus(String applicationId, String status, {String? offerCtc}) async {
    await _dio.patch(
      '/api/v1/placement/applications/$applicationId/status',
      data: {
        'status': status,
        'offer_ctc': offerCtc,
      },
    );
  }

  Future<void> scheduleInterview({
    required String applicationId,
    required String roundName,
    required String scheduledAt,
    String? locationOrLink,
    String? notes,
  }) async {
    await _dio.post(
      '/api/v1/placement/interviews/schedule',
      data: {
        'application_id': applicationId,
        'round_name': roundName,
        'scheduled_at': scheduledAt,
        'location_or_link': locationOrLink,
        'notes': notes,
      },
    );
  }

  Future<void> respondToOffer(String applicationId, String offerStatus) async {
    await _dio.patch(
      '/api/v1/placement/applications/$applicationId/offer-response',
      data: {'offer_status': offerStatus},
    );
  }
}

final placementRepositoryProvider = Provider<PlacementRepository>((ref) {
  final dio = ref.watch(dioClientProvider);
  return PlacementRepository(dio);
});
